### ISSUE TITLE
Failed to save key on readonly filesystem

### ISSUE THREAD

**Issue by mnapoli** [symfony/symfony#29357]
**Symfony version(s) affected**: 4.1.4

**Description**  

I run Symfony on a readonly filesystem and I can see this in my logs:

```
[warning] Failed to save key "%5BSymfony%5CBundle%5CFrameworkBundle%5CController%5CRedirectController%5D%5B1%5D" (array)
[warning] Failed to save key "%5BSymfony%5CBundle%5CFrameworkBundle%5CController%5CRedirectController%23urlRedirectAction%5D%5B1%5D" (array)
```

The `urlRedirectAction` is interesting, it seems like the routing might be caching a redirect response?

Am I doing something wrong here? I though the cache could be generated once (before deploying) and be readonly.

**How to reproduce**  

I would say deploy Symfony (with its cache generated) on a readonly filesystem. I deployed Symfony 4 with [Bref](https://github.com/mnapoli/bref) and its Symfony integration on AWS Lambda for example.

**Possible Solution**  

**Additional context**

**Comment by nicolas-grekas** [comment]
For read-only fs, you need to enable APCu - it'll be used automatically and fix the issue.
See `AbstractAdapter::createSystemCache()`

**Comment by mnapoli** [comment]
Oh OK thanks for the answer! Now what if I don't want to use APCu? ^^

On AWS lambda the filesystem is readonly (well, except `/tmp` so I could work around that) and APCu doesn't make sense because there is no shared memory as there is no webserver or PHP-FPM running. I don't even compile APCu there.

Also I'm curious: what is this cache about? Can I warm it up before deployment? Is it really important or can I disable it? (if you have a pointer to a documentation or something maybe?)

**Comment by nicolas-grekas** [comment]
To warm up the annotation cache, you need to dump the optimized composer class map then run cache:clear.
Then, you'll need to debug to answer your questions :)
Note that the warning are silent - you can also ignore them - it's just a cache that's missing and if that's not degrading perfs, why care?

**Comment by mnapoli** [comment]
> To warm up the annotation cache, you need to dump the optimized composer class map then run cache:clear.

Right, I'm already [running cache:clear and warmup before deploying](https://github.com/mnapoli/mnapoli.github.io/blob/master/.bref.yml#L3-L4) so that isn't it AFAICT.

BTW you mention the annotation cache but I don't think it's related to annotations. It mentions the `RedirectController`, the routing config is the only place I reference it: https://github.com/mnapoli/mnapoli.github.io/blob/925986f74f660cdb99b4c66f55cb37deaa73ae33/config/routes.yaml#L19 Is there a routing cache maybe?

> Then, you'll need to debug to answer your questions :)

Fair enough!

> Note that the warning are silent - you can also ignore them - it's just a cache that's missing and if that's not degrading perfs, why care?

This is not just one project (in this case my blog so indeed I don't care so much). But I'm trying to provide a proper open source integration [for Symfony on AWS Lambda](https://github.com/mnapoli/bref/blob/master/docs/Symfony.md). I want it to be as good as it can be. A warning tells me there's maybe something I'm missing and can be fixed.

Would it make sense to disable these warnings? (is that possible?)

**Comment by nicolas-grekas** [comment]
Disabling would be possible by injecting a NullLogger, but I wouldn't recommend that as it might remove useful info. In prod these should just go to /dev/null instead.

**Comment by jenschude** [comment]
I could fix it by overriding the annotation cache warmer definition with this one:

```
    annotations.cache_warmer:
        class: Symfony\Bundle\FrameworkBundle\CacheWarmer\AnnotationsCacheWarmer
        arguments:
            - '@annotations.reader'
            - '%kernel.cache_dir%/annotations.php'
            - '#^Symfony\\(?:Component\\HttpKernel\\|Bundle\\FrameworkBundle\\Controller\\(?!AbstractController$|RedirectController$|Controller$))#'
            - '%kernel.debug%'
```

Important here is the changed excludeRegExp which adds the RedirectController to be not excluded from the annotation cache.

**Comment by nicolas-grekas** [comment]
@jenschude would you mind submitting this as a PR?

**Comment by jenschude** [comment]
Was already on it. Only question for me is how to test this as there was no test for this before ;)

**Comment by nicolas-grekas** [comment]
Dunno :) That's ok not adding one for this case if it's too hard IMHO.
